core/trino-main/src/main/java/io/trino/execution/scheduler/BinPackingNodeAllocatorService.java

```diff
@@ -0,0 +1,533 @@
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static com.google.common.base.Verify.verify;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.util.concurrent.Futures.immediateFuture;
+import static io.airlift.concurrent.MoreFutures.getFutureValue;
+import static io.airlift.concurrent.Threads.daemonThreadsNamed;
+import static io.trino.spi.StandardErrorCode.NO_NODES_AVAILABLE;
+import static java.lang.Thread.currentThread;
+import static java.util.Comparator.comparing;
+import static java.util.Objects.requireNonNull;
+import static java.util.function.Predicate.not;
+
+public class BinPackingNodeAllocatorService
+        implements NodeAllocatorService
+{
+    private static final Logger log = Logger.get(BinPackingNodeAllocatorService.class);
+
+    @VisibleForTesting
+    static final int PROCESS_PENDING_ACQUIRES_DELAY_SECONDS = 5;
+    private static final int MAX_TASKS_PER_NODE = 1; // TODO be smart here (memory based)
+    private static final int MAX_FULL_NODES_PER_QUERY = 3;
+
+    private final NodeScheduler nodeScheduler;
+    private final int maxTasksPerNode;
+    private final int maxFullNodesPerQuery;
+
+    private final List<PendingAcquire> sharedPendingAcquires = new LinkedList<>();
+    private final Map<InternalNode, PendingAcquire> fullNodePendingAcquires = new HashMap<>();
+    private final Deque<PendingAcquire> detachedFullNodePendingAcquires = new ArrayDeque<>();
+
+    private final ConcurrentMap<InternalNode, Integer> sharedAllocationCounters = new ConcurrentHashMap<>();
+    private final Set<InternalNode> allocatedFullNodes = new HashSet<>();
+
+    private final Multimap<QueryId, InternalNode> fullNodesByQueryId = HashMultimap.create(); // both assigned pending and allocated
+
+    private final ScheduledThreadPoolExecutor executor = new ScheduledThreadPoolExecutor(1, daemonThreadsNamed("bin-packing-node-allocator"));
+    private final AtomicBoolean started = new AtomicBoolean();
+    private final AtomicBoolean stopped = new AtomicBoolean();
+    private final Semaphore processSemaphore = new Semaphore(0);
+
+    @Inject
+    public BinPackingNodeAllocatorService(NodeScheduler nodeScheduler)
+    {
+        this(nodeScheduler, MAX_TASKS_PER_NODE, MAX_FULL_NODES_PER_QUERY);
+    }
+
+    @VisibleForTesting
+    BinPackingNodeAllocatorService(NodeScheduler nodeScheduler, int maxTasksPerNode, int maxFullNodesPerQuery)
+    {
+        this.nodeScheduler = requireNonNull(nodeScheduler, "nodeScheduler is null");
+        this.maxTasksPerNode = maxTasksPerNode;
+        this.maxFullNodesPerQuery = maxFullNodesPerQuery;
+    }
+
+    @PostConstruct
+    public void start()
+    {
+        if (started.compareAndSet(false, true)) {
+            executor.schedule(() -> {
+                while (!stopped.get()) {
+                    try {
+                        processSemaphore.tryAcquire(PROCESS_PENDING_ACQUIRES_DELAY_SECONDS, TimeUnit.SECONDS);
```

Comment:
Please add a comment why is it needed to wake up periodically (to update nodes list)